corehq/motech/models.py

```diff
@@ -68,23 +93,103 @@ def plaintext_client_secret(self, plaintext):
             ciphertext=b64_aes_encrypt(plaintext)
             self.client_secret = f'${ALGO_AES}${ciphertext}'
 
+    @property
+    def last_token(self) -> Optional[dict]:
```

Comment:
Have you considered calling this `plaintext_last_token` to follow the pattern of the other encrypted secrets?

Also, that pattern seems ripe for abstraction, but I don't have any great ideas for how to do so - maybe a special `Field` class?  decorators?

Reply:
:thinking: I think this is a really good call. What do you think of something like this?

```python
class EncryptedMixin:

    def __init__(self, *args, *kwargs):
        self.placeholder = kwargs.pop('placeholder', None)
        super.__init__(*args, *kwargs)

    def to_python(self, value):
        if value.startswith(f'${ALGO_AES}$'):
            ciphertext = value.split('$', 2)[2]
            return b64_aes_decrypt(ciphertext)
        return value

    def get_db_prep_save(self, value, connection):
        value = self.get_db_prep_value(value, connection=connection,
                                       prepared=False)
        if self.placeholder is None or value != self.placeholder:
            ciphertext = b64_aes_encrypt(value)
            value = f'${ALGO_AES}${ciphertext}'
        return super().get_db_prep_save(value, connection)
```

Then I could use it to define `EncryptedCharField` and `EncryptedTextField`, and drop the "plaintext" properties. They would be plaintext in Python and encrypted in the DB.

(I'd prefer to do this in a separate branch.)

Reply:
Yeah, that looks great.  Are those two methods called automatically by Django when going from python to db and vice versa?  Might also be worth doing a quick search to be sure this doesn't already exist somewhere.